refactor(http): prepare for `noUncheckedIndexedAccess`

gabelluardo opened the pull request:
Ref: #4040

iuioiua commented:
Hi Gabriele, thank you for the PR. PTAL at https://github.com/denoland/deno_std/pull/4275#pullrequestreview-1861556153. It details how these PRs for #4040 should be approached.

gabelluardo commented:
Hi @iuioiua, thanks for the suggestion, I think I did it better now.

However, I have a doubt about these two functions:

https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L54-L69

and 

https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L71-L86

I tested with unit tests and seems theses two function can be replaced with `accept.split(",");` and `str.split(";").map((p) => p.trim())`. Is a test bug or we can remove the `for` loops?

iuioiua commented on http/_negotiation/encoding.ts:
```diff
@@ -158,5 +158,5 @@ export function preferredEncodings(
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
 }
```
Can you please justify this non-null assertion?

gabelluardo replied:
`priorities` is generated from the values of `provided` so `provided[priorities.indexOf(priority)]` is always non-null

iuioiua commented on http/_negotiation/language.ts:
```diff
@@ -140,5 +144,5 @@ export function preferredLanguages(
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
```
Ditto. Please justify this non-null assertion.

gabelluardo replied:
Same here

iuioiua commented on http/_negotiation/media_type.ts:
```diff
@@ -233,5 +236,5 @@ export function preferredMediaTypes(
     .filter(isQuality)
     .sort(compareSpecs)
-    .map((priority) => provided[priorities.indexOf(priority)]);
+    .map((priority) => provided[priorities.indexOf(priority)]!);
```
Ditto.

gabelluardo replied:
And here

iuioiua commented:
> Hi @iuioiua, thanks for the suggestion, I think I did it better now.
> 
> However, I have a doubt about these two functions:
> 
> https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L54-L69
> 
> and
> 
> https://github.com/denoland/deno_std/blob/083a2b12ba45381c78dbabfcf06d48e2da27f16b/http/_negotiation/media_type.ts#L71-L86
> 
> I tested with unit tests and seems theses two function can be replaced with `accept.split(",");` and `str.split(";").map((p) => p.trim())`. Is a test bug or we can remove the `for` loops?

Yeah, you're mostly right. Each of these can be replaced by `header.split(delimiter).map((p) => p.trim())`

iuioiua commented on http/server_test.ts:
```diff
@@ -155,7 +155,7 @@ Deno.test(
     try {
       assertEquals(server.addrs.length, 3);
-      assertEquals(server.addrs[0].transport, "tcp");
+      assertEquals((server.addrs[0] as Deno.NetAddr).transport, "tcp");
```
We can use non-null assertions here.

gabelluardo replied:
For the other parameters is used the casting, so I have to modify them too? Just for code coherence

iuioiua commented on http/server_test.ts:
```diff
@@ -164,7 +164,7 @@ Deno.test(
         listenerOneOptions.port,
       );
-      assertEquals(server.addrs[1].transport, "tcp");
+      assertEquals((server.addrs[1] as Deno.NetAddr).transport, "tcp");
```
Ditto

iuioiua commented on http/server_test.ts:
```diff
@@ -173,7 +173,7 @@ Deno.test(
         listenerTwoOptions.port,
       );
-      assertEquals(server.addrs[2].transport, "tcp");
+      assertEquals((server.addrs[2] as Deno.NetAddr).transport, "tcp");
```
Ditto

iuioiua commented on http/server_test.ts:
```diff
@@ -973,14 +973,14 @@ Deno.test(
     for (let i = 0; i < rejectionCount; i++) {
       assertEquals(
-        listener.acceptCallIntervals[i] >= expectedBackoffDelays[i],
+        listener.acceptCallIntervals[i]! >= expectedBackoffDelays[i]!,
```
Can you please justify these non-null assertions?

gabelluardo replied:
`rejectionCount` is the length of `expectedBackoffDelays` and both are not modified during the execution. The length of `listener.acceptCallIntervals` is `rejectionCount` by the initialization

iuioiua reviewed (approved):
LGTM. @kitsonk, please feel free to review as this code was introduced to the codebase by you.
